corehq/apps/app_manager/suite_xml/sections/endpoints.py

```diff
@@ -0,0 +1,69 @@
+from typing import List
+
+from django.utils.functional import cached_property
+
+from corehq.apps.app_manager import id_strings
+from corehq.apps.app_manager.suite_xml.contributors import (
+    SuiteContributorByModule,
+)
+from corehq.apps.app_manager.suite_xml.post_process.workflow import (
+    WorkflowDatumMeta,
+    WorkflowHelper,
+)
+from corehq.apps.app_manager.suite_xml.xml_models import (
+    Argument,
+    PushFrame,
+    SessionEndpoint,
+    Stack,
+    StackDatum,
+)
+from corehq.apps.app_manager.xpath import XPath
+
+
+class SessionEndpointContributor(SuiteContributorByModule):
+    """
+    Generates "Session Endpoints" - user-defined labels for forms or modules.
+    They end up as entries in the suite file that declare stack operations
+    necessary to navigate to the form or module, as well as what arguments (eg:
+    case IDs) must be provided to get there.
+    """
+
+    def get_module_contributions(self, module) -> List[SessionEndpoint]:
+        endpoints = []
+        if module.session_endpoint_id:
+            endpoints.append(self._get_module_endpoint(module))
+        for form in module.get_suite_forms():
+            if form.session_endpoint_id:
+                endpoints.append(self._get_form_endpoint(form, module))
+        return endpoints
+
+    def _get_module_endpoint(self, module):
+        id_string = id_strings.case_list_command(module)
+        return self._make_session_endpoint(id_string, module, module.session_endpoint_id)
+
+    def _get_form_endpoint(self, form, module):
+        id_string = id_strings.form_command(form)
+        return self._make_session_endpoint(id_string, module, form.session_endpoint_id)
+
+    @cached_property
+    def _workflow_helper(self):
+        return WorkflowHelper(self.suite, self.app, self.modules)
+
+    def _make_session_endpoint(self, id_string, module, endpoint_id):
+        stack = Stack()
+        frame = PushFrame()
+        stack.add_frame(frame)
+        frame.add_command(XPath.string(id_string))
+        arguments = []
+        for child in self._workflow_helper.get_frame_children(id_string, module):
+            if isinstance(child, WorkflowDatumMeta):
+                arguments.append(Argument(id=child.id))
+                frame.add_datum(
+                    StackDatum(id=child.id, value=f"${child.id}")
```

Comment:
Is it always obvious to the user what this id will be? I'm assuming for advanced modules that this will be something like `case_id_new_{case type}_{action index}`. Is that correct, or have I misunderstood what this is actually doing?

Reply:
I'm actually not sure what happens on the other side of this equation.  The reminders app (or other future usage of this feature) _will_ need to know how to fill out this information, but I'm not certain how they do that.  It should be the same as how these data are made available in the session.  Do you have any thoughts around how to make it clearer?

Reply:
Got it. I guess if you are using advanced modules you are on your own anyway, and the format for the case ids are pretty well documented. I had been under the impression that the argument ID was something that the user would configure (so they would remain static if anything else about the app changes), but sounds like this happens automatically. I don't understand this feature well enough to know if that is an issue or not, so can leave it up to you.

Reply:
Ah, gotcha - I think you're thinking of the endpoint ID.  That's an attribute of the `<endpoint />` node wrapping this.  In an earlier version of this PR, that was also used as the argument ID, but Shubham pointed out the argument ID should match the datum ID.  Honestly I don't have my head wrapped around exactly how that's used, but my understanding is the argument ID is to be used by the other app to figure out how to fill out the stack.